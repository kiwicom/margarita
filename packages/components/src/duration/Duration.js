// @flow

import * as React from 'react';
import { View } from 'react-native';
import {
  Text,
  StyleSheet,
  Icon,
  type StylePropType,
} from '@kiwicom/universal-components';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

type Props = {|
  +showIcon: boolean,
  +duration: ?number,
  +style?: StylePropType,
  +iconStyle?: StylePropType,
|};

export const separateHours = (durationInMinutes: number) => {
  let minutes = durationInMinutes;

  if (minutes < 0) {
    minutes = 0;
  }

  const hours = Math.floor(minutes / 60);
  minutes %= 60;

  return { hours, minutes };
};

export default function Duration(props: Props) {
  const { hours, minutes } = separateHours(props.duration || 0);
  const duration = `${hours}h ${minutes}m`;
  return (
    <View style={styleSheet.row}>
      {props.showIcon && (
        <Icon
          name="clock"
          size="small"
          color={defaultTokens.colorTextSecondary}
          style={props.iconStyle}
        />
      )}
      <Text style={[styleSheet.durationText, props.style]}>{duration}</Text>
    </View>
  );
}

Duration.defaultProps = {
  showIcon: true,
};

const styleSheet = StyleSheet.create({
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  durationText: {
    fontSize: 12,
    color: defaultTokens.colorTextSecondary,
  },
});
