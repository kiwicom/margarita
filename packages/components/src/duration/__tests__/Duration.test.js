// @flow

import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import Duration, { separateHours } from '../Duration';

const renderer = new ShallowRenderer();

it('works as expected', () => {
  const test = (duration, expected) => {
    expect(separateHours(duration)).toEqual(expected);
  };

  test(-1, { hours: 0, minutes: 0 });
  test(0, { hours: 0, minutes: 0 });
  test(1, { hours: 0, minutes: 1 });
  test(59, { hours: 0, minutes: 59 });
  test(60, { hours: 1, minutes: 0 });
  test(121, { hours: 2, minutes: 1 });
  test(10080, { hours: 168, minutes: 0 });
});

it('works with broken input', () => {
  expect(renderer.render(<Duration duration={undefined} />)).toMatchSnapshot();

  expect(renderer.render(<Duration duration={null} />)).toMatchSnapshot();
});
