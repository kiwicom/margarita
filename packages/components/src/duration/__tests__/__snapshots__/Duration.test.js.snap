// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works with broken input 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <Icon
    color="#7f91a8"
    name="clock"
    size="small"
  />
  <Text
    style={
      Array [
        Object {
          "color": "#7f91a8",
          "fontSize": 12,
        },
        undefined,
      ]
    }
  >
    0h 0m
  </Text>
</Component>
`;

exports[`works with broken input 2`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <Icon
    color="#7f91a8"
    name="clock"
    size="small"
  />
  <Text
    style={
      Array [
        Object {
          "color": "#7f91a8",
          "fontSize": 12,
        },
        undefined,
      ]
    }
  >
    0h 0m
  </Text>
</Component>
`;
